src/main/java/jenkins/plugins/openstack/compute/SlaveOptions.java

```diff
@@ -229,11 +248,16 @@ private Object readResolve() {
                 .jvmOptions(_override(this.jvmOptions, o.jvmOptions))
                 .fsRoot(_override(this.fsRoot, o.fsRoot))
                 .launcherFactory(_override(this.launcherFactory, o.launcherFactory))
+                .nodeProperties(_overrideC(this.nodeProperties, o.nodeProperties))
                 .retentionTime(_override(this.retentionTime, o.retentionTime))
                 .build()
         ;
     }
 
+    private @CheckForNull <T extends Collection<?>> T _overrideC(@CheckForNull T base, @CheckForNull T override) {
+        return override == null || override.isEmpty() ? base : override;
```

Comment:
I suspect the need for specific override for collections is a symptom of a problematic contract of `SlaveOptions#nodeProperties`. There is a need to signal emptiness but we should have only one value for it (empty collection or `null`). I lean towards using `null` and make sure the collection is never empty.

Reply:
Actually, scratch that. Empty collection is a valid value for overriding slave options. Ex.: Default options specify some properties and template can override by making the collection empty (effectively removing the options passed in as defaults). Empty collection should be handled as a legitimate value here.

Reply:
I mostly agree ... but I'm not sure how, via the configuration WebUI, the user can distinguish between an empty list and a null list.  Unless we can let the user distinguish between the two cases, I think that the plugin functionality should treat them the same.

The jelly code is in [SlaveOptions/config.jelly line 68](https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/files#diff-b4cd7ab6e132cf606fb8931efff7fd35R68) and, as far as I am aware, this gives the user no control over whether it results in an empty list or a null list.

If you can help (or get someone else to help...) work out the correct jelly magic that'll give the user control over null/empty then I'd be happy to remove the "treat empty the same as null" code.

Reply:
Good point. I am afraid we have to enclose that with `optionalBlock` that, when unchecked would mean _I do not contribute/override any options_, and when checked it would be an explicit list of options to provide/override. That would enable the semantics of both `null` (checkbox unchecked) and empty (checked but no implementations added).

---

I do not quite like the complexity but I see no other way to permits overriding templates to erase inherited defaults.

Reply:
Can you point me at an example of where this is done in existing code?
I'm familiar with the _java_ `Optional` but doing this with Jenkins jelly code is unknown territory for me...

Reply:
I've been googling...  from what I can tell, the jelly `optionalBlock` is controlled by a boolean, which I guess we'd call `hasNodeProperties`.
So, I figure there's three possibilities:
1. (worst case) We need to _store_ that boolean, i.e. we add a new boolean field `hasNodeProperties`, together with all the getters/setters that entails, and then ensure that getNodeProperties returns null if hasNodeProperties is false.
1. As above, but we can get away with it being a transient boolean whose value is initialised as nodeProperties!=null.
1. (best case) We just implement getHasNodeProperties and setHasNodeProperties to give the jelly something to talk to ... and rely on setHasNodeProperties always being called before setNodeProperties.

WDYT?

Reply:
I would have to play with that to see whether that is implementable without too much of a hassle, but we can pass the optional block an explicit condition of `checked="${it.nodeProperties != null}"` so there should be no need to maintain the boolean property alongside of the properties list. (The submitted input value will be just ignored by the `@DataBoundConstructor`).

Let me know if you prefer me to prototype that.

Reply:
All help would be gratefully accepted :grinning: 
FYI I only implemented this because I wanted https://github.com/jenkinsci/wsclean-plugin to have a "don't clean this node" nodeProperty - what was meant to be a "fairly quick" enhancement to wsclean-plugin has snowballed out of control as I had to also add NodeProperty support to the vSphere plugin as well as this plugin (vSphere was easy compared to this one though).

Reply:
I've included your changes (olivergondza@e2ab58a) in https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/commits/5456d1d835b8da5588b9f7e3ce96730cc0b1a071
I've done some basic testing locally and it looks like it works exactly as expected :grin: